src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
+        return table;
     }
 
     public void put(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null) {
-                v = new Vector<V>(ies);
-                table.put(key, v);
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>();
             }
-            v.addElement(value);
-        }
+            list.add(value);
+            return list;
+        });
+    }
+
+    public void putAll(K key, Collection<V> elements) {
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>(elements);
+            } else {
+                list.addAll(elements);
+            }
+            return list;
+        });
     }
 
     /**
      * Returns the first element for this key.
      */
     public V get(K key) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v == null ?
-                    null :
-                    v.firstElement());
+        return getFirst(key);
+    }
+
+    /**
+     * Returns the first element for this key.
+     */
+    public V getFirst(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return null;
         }
+        return list.get(0);
     }
 
     public boolean containsKey(K key) {
-		synchronized (table) {
-			return table.containsKey(key);
-		}
+        return this.table.containsKey(key);
     }
 
     public boolean containsElement(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v != null) && v.contains(value);
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return false;
         }
+        return list.contains(value);
     }
 
     /**
-     * Users will have to handle synchronizing.
+     * Returns mapped value collection as {@link Enumeration}
+     *
+     * <b>Note:</b> this method is deprecated
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a {@link List}
+     *
+     * @param key key mapping
      */
+    @Deprecated
     public Enumeration<V> getAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-        return (v == null ?
-                new EmptyEnumeration<V>() :
-                v.elements());
+        List<V> elements = getAllAsList(key);
+        if (elements.isEmpty()) {
+            return Collections.emptyEnumeration();
+        } else {
+            return Collections.enumeration(elements);
+        }
+    }
+
+    public List<V> getAllAsList(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return Collections.unmodifiableList(list);
```

Comment:
This returns an unmodifiable view of the underlying list. This means that when a new entry is inserted or an existing entry deleted in the table for this key, the change is reflected in this returned list. That is highly unexpected (and if intended, should be thoroughly documented), most likely undesired, and probably  an unintended effect of choice of datastructure. See also my comment about the use of CoWList.

Reply:
This was also true for the `Enumeration` objects, which were returned before.
I think that nobody ever used this like that and wanted any updates during iteration or later use.
I agree that it is really better to return "safe" and stable collections and iterators.

Now, when lists are truly immutable, then all returned values are also not modifiable.
Please see the new changes, it is probably better now.